src/charm.py

```diff
@@ -261,38 +261,71 @@ def _get_gateway_apis_kubernetes_resource_manager(self):
             logger=LOGGER,
         )
 
-    def _workload_tracing_config(self) -> Dict[str, str]:
-        """Set workload tracing configuration."""
-        tracing_config = {}
-        if not self.workload_tracing.is_ready():
-            return tracing_config
-
-        workload_tracing_endpoint = self.workload_tracing.get_endpoint("otlp_grpc")
-        if not workload_tracing_endpoint:
-            return tracing_config
-
-        tracing_config["meshConfig.enableTracing"] = "true"
-        tracing_config["meshConfig.extensionProviders[0].name"] = "otel-tracing"
-        tracing_config["meshConfig.extensionProviders[0].opentelemetry.port"] = urlparse(
-            f"//{workload_tracing_endpoint}"
-        ).port
-        tracing_config["meshConfig.extensionProviders[0].opentelemetry.service"] = urlparse(
-            f"//{workload_tracing_endpoint}"
-        ).hostname
-        tracing_config["meshConfig.defaultProviders.tracing[0]"] = "otel-tracing"
-        tracing_config["meshConfig.defaultConfig.tracing.sampling"] = 100.0
-
-        return tracing_config
+    def _workload_tracing_provider(self) -> Tuple[List[Any], Dict[str, Any]]:
+        """Return a tuple with a list containing the tracing provider and global tracing settings."""
+        providers = []
+        global_config = {}
+        if self.workload_tracing.is_ready():
+            workload_tracing_endpoint = self.workload_tracing.get_endpoint("otlp_grpc")
+            if workload_tracing_endpoint:
+                parsed = urlparse(f"//{workload_tracing_endpoint}")
+                providers.append(
+                    {
+                        "name": "otel-tracing",
+                        "opentelemetry": {
+                            "port": str(parsed.port),
+                            "service": parsed.hostname,
+                        },
+                    }
+                )
+                global_config = {
+                    "meshConfig.enableTracing": "true",
+                    "meshConfig.defaultProviders.tracing": "otel-tracing",
+                    "meshConfig.defaultConfig.tracing.sampling": 100.0,
+                }
+        return providers, global_config
+
+    def _flatten_config(self, value: Any, prefix: str = "") -> Dict[str, Any]:
+        """Recursively flatten a nested dictionary or list into a dictionary of key/value pairs."""
+        flat: Dict[str, Any] = {}
+        if isinstance(value, dict):
+            for k, v in value.items():
+                new_prefix = f"{prefix}.{k}" if prefix else k
+                flat.update(self._flatten_config(v, new_prefix))
+        elif isinstance(value, list):
+            for i, item in enumerate(value):
+                new_prefix = f"{prefix}[{i}]"
+                flat.update(self._flatten_config(item, new_prefix))
+        else:
+            flat[prefix] = value
+        return flat
```

Comment:
I think there's a bug here, but I can't tell.  Some unit testing will prove it right or wrong.

Why I think there's a bug is that it does `flat[prefix] = value`.  That will fail if `prefix=""`.  For this particular use case i dont think you'll have `prefix=""`, but in general you could (`_flatten_config({"a": {"b": "c"}}, prefix="")` should be valid

This'll be really easy to figure out with a few unit tests though.  Make a basic test with a good set of parameters to catch the edge cases and you'll know for sure

Reply:
Write it however you like, but I think it'll be easier to write if the return is a list of (key, value) tuples rather than a dict.  But if what you have works and you like it better, lgtm :) 

The reason I think about the list of tuples is I think it removes the need to pass prefixes all the way down.  The calling function can just recurse with something like `flat.extend([(prefix + k, v) for k, v in flatten_config(value)`. 
 But if that makes things harder, ignore me :D  Just an idea in case it helps

Reply:
ty for tests, lgtm!  also, I never really thought about how `""` is a valid key for a dict...